barretenberg/cpp/src/barretenberg/commitment_schemes/commitment_key.hpp

```diff
@@ -81,6 +82,65 @@ template <class Curve> class CommitmentKey {
         return scalar_multiplication::pippenger_unsafe<Curve>(
             const_cast<Fr*>(polynomial.data()), srs->get_monomial_points(), degree, pippenger_runtime_state);
     };
+
+    /**
+     * @brief Efficiently commit to a sparse polynomial
+     * @details Iterate through the {point, scalar} pairs that define the inputs to the commitment MSM, maintain (copy)
+     * only those for which the scalar is nonzero, then perform the MSM on the reduced inputs.
+     * @warning Method makes a copy of all {point, scalar} pairs that comprise the reduced input. Will not be efficient
+     * in terms of memory or computation for polynomials beyond a certain sparseness threshold.
+     *
+     * @param polynomial
+     * @return Commitment
+     */
+    Commitment commit_sparse(std::span<const Fr> polynomial)
+    {
+        BB_OP_COUNT_TIME();
+        const size_t degree = polynomial.size();
+        ASSERT(degree <= srs->get_monomial_size());
+
+        // Extract the precomputed point table (contains raw SRS points at even indices and the corresponding
+        // endomorphism point (\beta*x, -y) at odd indices).
+        G1* point_table = srs->get_monomial_points();
+
+        // Define structures needed to multithread the extraction of non-zero inputs
+        const size_t num_threads = degree >= get_num_cpus_pow2() ? get_num_cpus_pow2() : 1;
```

Comment:
IIUC these shouldn't (aren't) the number of threads, just the number of things you need to iterate. If you look closely at the current implementation of `parallel_for`, the number of threads are not sth the caller can choose, you can only choose the number of elements to iterate (which is right): https://github.com/AztecProtocol/aztec-packages/blob/603559519e6882e94bb7071a9237bf12004fff2f/barretenberg/cpp/src/barretenberg/common/parallel_for_mutex_pool.cpp#L124

Reply:
I guess I'm a little unclear what your suggestion is here. Is your objection about the name `num_threads`? I guess I'm thinking of it as shorthand for "num threads over which to distribute the work", which is how this value is used if I'm not mistaken. Admittedly though I am being a bit sloppy here in that I don't really need a power-of-2 thread count and my condition for when not to multithread is a bit arbitrary. In practice though the only context that matters is degree ~2^18 and higher so not so important to get the small values right

Reply:
What I mean is that the value passed to `parallel_for` is in principle not the number of threads. If you have 16 cores and do `parallel_for(200, some_f)` then it will be chunked in 200/16 chunks over 16 threads, and f will be called from 0 to 199.

Maybe the way you set it up does make it coincide with the number of threads though. @ludamad would be the best to ask. I'm commenting because I had to use `parallel_for` recently and noticed this subtlety.

Reply:
ah I see what you mean. In my case the "num_iterations" input will always be <= actual num threads but your point stands. I don't really love that - seems like the parallel_for interface should allow you to specify how it should multithread.